### PR TITLE
[BEAM-177] Upload JaCoCo coverage data to Coveralls from Travis-CI

### PR DESCRIPTION
**Pull request by kennknowles**

### CODE REVIEW

**Comment by kennknowles** [comment]
Example result: https://coveralls.io/jobs/13374509

**Comment by kennknowles** [comment]
The important bit is browser integration into code review. For coveralls there seems to be [this](https://github.com/kwonoj/coveralls-overlay).

**Comment by kennknowles** [comment]
The failure is in the `RunnableOnService` tests, unrelated to the contents of the PR. The job run by the test seems to have a non-unique job name.

File: .travis.yml [hunk 1]
```diff
@@ -31,6 +31,9 @@ before_install:
 install:
   - travis_retry mvn -B install clean -U -DskipTests=true
 
+after_success:
```
Comment (davorbonaci) [inline .travis.yml]: `after_success` logically comes after `script`. Move below.

File: .travis.yml [hunk 2]
```diff
@@ -31,6 +31,9 @@ before_install:
...
+  - mvn jacoco:report coveralls:report
```
Comment (davorbonaci) [inline .travis.yml]: `mvn install` runs `jacoco:report`, I think. If so, `coveralls:report` might be all that is needed.

Even better fix would be to stop running jacoco:report as part of the mainline Maven build.

**Comment by davorbonaci** [comment]
R: @davorbonaci

Meta-comment: I'd like Jenkins to do this instead of Travis. Not sure what will happen if they both attempt to do it.

File: .travis.yml [hunk 3]
```diff
@@ -31,6 +31,9 @@ before_install:
...
+
```
Comment (davorbonaci) [inline .travis.yml]: Also, not sure what will happen when Travis tries to do this from its test matrix. We should probably make sure it runs just once, not 4 times.

Comment (kennknowles) [inline .travis.yml]: It runs just once, for the first job. Of course, coverage values could actually differ, but I think that is beyond coveralls.

**Comment by kennknowles** [comment]
Re: Travis vs Jenkins I don't care so much in this case. It should be fast, so it does not slow down code review waiting for results. If Jenkins is faster, then it is definitely the best choice.

But I have a different stance on the meta-issue:

1. Travis (or equivalent - even our own `.jenkins.sh`) piggybacks on our existing processes perfectly.
   - Process: Code review of the configuration changes. This implies higher quality results and more transparency about when and why changes occur.
   - Permissions: Anyone can propose a change in a PR, and any committer can accept it.
   - Tools: Text editor, git. And the UI, while less powerful overall, is easier for newcomers.
2. Travis has good integrations with lots of services. For example, you can use coveralls without a repo token :-)
3. Travis is trivial to configure.

Jenkins has capabilities that Travis does not have:

1. It has a nice test result UI. This is the only reason I can think of to run unit tests there.
2. You can configure a build that won't just obey the instructions in a PR, so it can do more sensitive things or be very expensive to run. (integration tests, submit queue)
3. Lots of plugins.
4. (Hopefully) lower latency for uploading coverage.

These are compelling. We should use Jenkins for these things. But I don't see a strong reason to use it for more than that, especially since there are major downsides - no source control of config, more complex config, having to issue tickets to infra for some changes, silent back-channel changes, no review of config changes, no presubmit of config changes, no ability to synchronously alter config and code, ...

For now, I'd be happy to merge PRs and get immediate benefit from Travis until our Jenkins config is offering equivalent capabilities. Perhaps it will be quite soon, but the cost is zero either way.

**Comment by kennknowles** [comment]
BTW examining the latency, Jenkins is much much lower latency so we should go with Jenkins for this. Travis in the meantime.

**Comment by kennknowles** [comment]
I should have remembered earlier that we can have both: https://wiki.jenkins-ci.org/display/JENKINS/Travis+YML+Plugin

File: .travis.yml (hunk 3 above)
Comment (kennknowles) [inline .travis.yml]: Explicitly running it just on Java 8 configuration, since that covers more.

File: .travis.yml (hunk 1 above)
Comment (kennknowles) [inline .travis.yml]: Done.

**Comment by kennknowles** [comment]
PTAL. Tests pass and we can start getting some benefit for all reviews that happen long enough after a PR shows up.

**Comment by davorbonaci** [comment]
Please test.

`[ERROR] Failed to execute goal org.eluder.coveralls:coveralls-maven-plugin:4.1.0:report (default-cli) on project parent: I/O operation failed: No coverage report files found -> [Help 1]`

**Comment by kennknowles** [comment]
The problem is this: Running the JaCoCo agent is incompatible with setting `forkCount=0`

**Comment by davorbonaci** [comment]
The same error still seems to occur.

**Comment by kennknowles** [comment]
Yes, I did not remove the setting of `forkCount=0`. Why was it added?

**Comment by davorbonaci** [comment]
I believe @dhalperi added it -- Travis build was failing without it at some point.

**Comment by davorbonaci** [comment]
* Jacoco doesn't work if `forkCount` is set to 0. See [here](http://eclemma.org/jacoco/trunk/doc/maven.html) for more details.
* We set `forkCount` to zero in the Travis environment to control memory usage. Empirically, if we set it to one, it will fail with out-of-memory issue almost always.

Basically, to move forward, we'd have to investigate memory usage of individual tests and get it down. Then, we can increase `forkCount, and finally enable Jacoco in Travis.

Given that it "somewhat works" in Jenkins right now, that Jenkins is generally preferable over Travis, that it seems easier to fully fix Jenkins than Travis, etc., I'd just probably close this pull request.

**Comment by kennknowles** [comment]
SGTM. Probably a red flag that we blow memory running unit tests, though.

**Comment by dhalperi** [comment]
It's actually simpler than that. Since Maven does not copy memory limits from the root process to the forked processes, Java never garbage collects (since it does not think it needs to). The VM happily grows the (non-leaking) memory until Travis kills it well before the Java memory limit is reached.

**Comment by kennknowles** [comment]
I see. This can be fixed via maven config in a few ways I think. I suppose it may not be worth making the pom more complex.

**Pull request closed without merging**